### PR TITLE
Feature/config

### PR DESCRIPTION
**Pull request by hdonnay**
This is a rework of the config package to have it split into a distinct module with minimal dependencies.
I added a linter because it stuck in my mind after I realized it could be done once all the types are in the same package.

This required pretty extensive surgery to the notifier's types, as the config referenced those directly.

Outstanding questions before the final version:

- Should we break the config package code API more? 
  Namely, rework the `Validate` method. `Lint` could become a replacement for `Validate` and return errors in same/similar situations. Perhaps unexport the `Mode` member and add setter/getters for it.
- How can we evolve defaults?
  Is that even possible/desirable when a user renders out a config and feeds it back?

### CODE REVIEW

File: cmd/clair/values.go [hunk 1]
```diff
@@ -37,55 +37,33 @@ func (v *ConfValue) Set(s string) error {
...
-		*v = ModeCombo
-	case "index", "indexer":
-		*v = ModeIndexer
-	case "match", "matcher":
-		*v = ModeMatcher
-	case "notify", "notifier":
-		*v = ModeNotifier
-	default:
-		return fmt.Errorf("unknown mode argument %q", s)
+	case "", "dev", "pizza", "combination":
```
Comment (crozzy) [inline cmd/clair/values.go]: you don't wanna keep combo?

File: cmd/clair/values.go [hunk 2]
```diff
@@ -37,55 +37,33 @@ func (v *ConfValue) Set(s string) error {
...
+		// "Pizza", of course, being the best Combos flavor.
+		s = "combo"
+	case "index":
+		s = "indexer"
+	case "match":
+		s = "matcher"
+	case "notify":
+		s = "notifier"
```
Comment (crozzy) [inline cmd/clair/values.go]: I kind of think we should just assign the mode here and this effectively becomes `ParseMode`, I get the case-insensitivity, but we're going to document one way, so adding another way might confuse things (I do enjoy pizza though).

File: config/tls.go [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+	// The filesystem path where a root CA can be read.
+	//
+	// This can also be controlled by the SSL_CERT_FILE and SSL_CERT_DIR
+	// environment variables, or adding the relevant certs to the system trust
+	// store.
+	RootCA string `yaml:"root_ca" json:"root_ca"`
+	// The filesystem path where a tls certificate can be read.
+	Cert string `yaml:"cert" json:"cert"`
+	// The filesystem path where a tls private key can be read.
+	Key string `yaml:"key" json:"key"`
```
Comment (crozzy) [inline config/tls.go:26]: I think maybe these should be suffixed with `_path` or `_file`

File: notifier/amqp/config.go [hunk 4]
```diff
@@ -3,135 +3,80 @@ package amqp
...
+	AutoDelete bool
 }
 
-// Validate confirms configuration is valid and fills in private members
-// with parsed values on success.
-func (c *Config) Validate() (Config, error) {
-	conf := *c
-	if c.Exchange.Type == "" {
-		return conf, fmt.Errorf("AMQP config requires the exchange.type field")
+func loadTLSConfig(c *config.AMQP) (*tls.Config, error) {
```
Comment (crozzy) [inline notifier/amqp/config.go]: What are your thoughts on `loadTLSConfig` being generic for all messaging backends and the check to see whether to load TLS is done before calling this function. Just thinking we might want to extend this functionality to layer fetching and potentially vulnerability fetching. Just a thought

**Review by crozzy — COMMENTED** [review]
Wow this got deep, I know it's in draft was just interested in taking a look. I'll review again when finished

File: config/tls.go (hunk 3 above)
Comment (hdonnay) [reply config/tls.go:26]: I think we have to leave it until a v2, as it'll change the on-disk representation.

File: cmd/clair/values.go (hunk 2 above)
Comment (hdonnay) [reply cmd/clair/values.go]: My preference, then, would be to just delegate to the `ParseMode` function.

File: cmd/clair/values.go (hunk 1 above)
Comment (hdonnay) [reply cmd/clair/values.go]: it's still in there, just implicitly via the ParseMode call

**Comment by hdonnay** [comment]
For my questions:

- I broke the API a bit more. I think the better output around potential issues with the config is worth it.
- Punting on the defaults question. I think we have to accept that if you render out a config object, it has the defaults from that point in time.

File: cmd/clairctl/export.go [hunk 5]
```diff
@@ -62,11 +63,18 @@ func exportAction(c *cli.Context) error {
 	}
 	cfgs := make(map[string]driver.ConfigUnmarshaler, len(cfg.Updaters.Config))
 	for name, node := range cfg.Updaters.Config {
-		cfgs[name] = node.Decode
+		node := node // Unsure if needed.
```
Comment (crozzy) [inline cmd/clairctl/export.go]: Difficult one, I believe yes as when the function is evaluated `n` will already be at it's last value unless copied https://play.golang.org/p/74vXJ_Xx2qI

File: config/config.go [hunk 6]
```diff
@@ -2,32 +2,19 @@ package config
...
-
+// Config is the configuration object for the commands in
+// github.com/quay/clair/v4/cmd/...
 type Config struct {
 	// One of the following strings
 	// Sets which mode the clair instances will run in
 	//
 	// "indexer": runs just the indexer node
 	// "matcher": runs just the matcher node
 	// "combo":	will run both indexer and matcher on the same node.
```
Comment (crozzy) [inline config/config.go]: Poor notifier, always the black sheep

File: config/defaults.go [hunk 7]
```diff
@@ -0,0 +1,29 @@
...
+	// DefaultAddress is used if an "http_listen_addr" is not provided in the config.
+	DefaultAddress = ":6060"
+	// DefaultScanLockRetry is the default retry period for attempting locks
+	// during the indexing process. Its name is a historical accident.
+	DefaultScanLockRetry = 1
+	// DefaultMatcherPeriod is the default interval for running updaters.
+	DefaultMatcherPeriod = 30 * time.Minute
+	// DefaultMatcherRetention is the number of updates per vulnerability
+	// database to retain.
+	DefaultMatcherRetention = 10
```
Comment (crozzy) [inline config/defaults.go]: Thoughts on this name, I think `DefaultUpdateRetention` might be closer to the actual function.

File: notifier/amqp/config.go [hunk 8]
```diff
@@ -1,137 +1,33 @@
...
-		conf.tls = &TLS
-	}
-
-	if !c.Direct {
-		callback, err := url.Parse(c.Callback)
-		if err != nil {
-			return conf, fmt.Errorf("failed to parse callback url")
-		}
-		conf.callback = *callback
+func exchangeFrom(c *config.AMQP) exchange {
```
Comment (crozzy) [inline notifier/amqp/config.go]: What is the point of doing this and not assigning directly?
```go
d.exchange = conf.Exchange
```

File: notifier/processor_create_test.go [hunk 9]
```diff
@@ -79,7 +80,7 @@ func TestProcessCreate(t *testing.T) {
 // available
 func testProcessorStoreErr(t *testing.T) {
 	t.Parallel()
-	ctx := context.TODO()
+	ctx := zlog.Test(context.Background(), t)
```
Comment (crozzy) [inline notifier/processor_create_test.go:83]: Kinda neat idea

File: notifier/stomp/config.go [hunk 10]
```diff
@@ -3,95 +3,44 @@ package stomp
...
+)
 
-	if !c.Direct {
-		var u *url.URL
-		var err error
-		if u, err = url.Parse(c.Callback); err != nil {
-			return conf, fmt.Errorf("direct delivery is disabled but callback url could not be parsed.")
-		}
-		conf.callback = *u
+func loadTLSConfig(c *config.STOMP) (*tls.Config, error) {
```
Comment (crozzy) [inline notifier/stomp/config.go]: I can't tell where this is being called now

**Review by crozzy — CHANGES_REQUESTED** [review]
Really great improvement, just a few comments

File: notifier/stomp/config.go (hunk 10 above)
Comment (hdonnay) [reply notifier/stomp/config.go]: Might be left over, I'll double-check

File: notifier/amqp/config.go (hunk 8 above)
Comment (hdonnay) [reply notifier/amqp/config.go]: This way, nothing's holding onto pointers back to the config object.

Comment (crozzy) [reply notifier/amqp/config.go]: Totally agree with the intent but as I see it `Exchange` in the config isn't passed by it's reference, so assigning it directly to a `Deliverer.Exchange` field would mean the values are copied over to a new object. Any mutation of `config.Exchange` or `Deliverer.Exchange` should not affect each other I believe.

Comment (crozzy) [reply notifier/amqp/config.go]: (leaving aside the whole "nothing is passed by reference" debate)

**Review by crozzy — APPROVED** [review]
LGTM!

**Pull request merged**